### PR TITLE
Add `dataloader_idx` to batch transfer hooks

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/6058
Also updated the internal flow from args to kwargs (OrderedDict) to simplify things a bit and better accessibility and kwargs values are passed as args to the user-specific hooks for BC.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -184,7 +184,9 @@ def logger(self):
         """ Reference to the logger object in the Trainer. """
         return self.trainer.logger if self.trainer else None
 
-    def _apply_batch_transfer_handler(self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0):
+    def _apply_batch_transfer_handler(
+        self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: Optional[int] = None
+    ):
         batch = self.on_before_batch_transfer(batch, dataloader_idx)
         batch = self.transfer_batch_to_device(batch, device)
```
Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:191]: I think we should remove the `device` argument from `transfer_batch_to_device` since this can easily be accessed using `self.device` (device property of LightningModule).

**Comment by rohitgr7** [comment]
Hi @SeanNaren some deepspeed/sharded tests are failing here. Mind check the reason for it. Can't figure out what's causing these failures. Thanks :)

**Comment by rohitgr7** [comment]
not sure why `test_datamodule_parameter` test is failing in some checks. Working fine for me locally.

File: pytorch_lightning/accelerators/accelerator.py [hunk 2]
```diff
@@ -158,12 +161,12 @@ def on_train_start(self) -> None:
 
     def training_step(
         self,
-        args: List[Union[Any, int]],
+        kwargs: Dict[str, Union[Any, int]],
```
Comment (tchaton) [inline pytorch_lightning/accelerators/accelerator.py:164]: Does this work when the batch is a dict ?

Comment (rohitgr7) [reply pytorch_lightning/accelerators/accelerator.py:164]: yes, it does. Everything is the same, it's just that previously it was a tuple (batch, batch_idx, ...) and not it is an OrderedDict {'batch': batch, 'batch_idx': batch_idx}.

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -834,9 +835,9 @@ def should_check_val_fx(self, batch_idx, is_last_batch, on_epoch=False):
 
         return should_check_val and can_check_val
 
-    def build_train_args(self, batch, batch_idx, opt_idx, hiddens):
+    def _build_kwargs(self, batch, batch_idx, opt_idx, hiddens):
         # enable not needing to add opt_idx to training_step
-        args = [batch, batch_idx]
+        kwargs = OrderedDict({'batch': batch, 'batch_idx': batch_idx})
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:840]: This does not work as the moment you build the inner dict you lose the ordered guarantee. You need to use a list of tuples

File: tests/trainer/test_dataloaders.py [hunk 4]
```diff
@@ -1159,3 +1159,72 @@ def test_replace_sampler_with_multiprocessing_context(tmpdir):
...
     assert (new_data_loader.multiprocessing_context == train.multiprocessing_context)
+
+
+@pytest.mark.parametrize('multiple_trainloader_mode', ["min_size", "max_size_cycle"])
+def test_correct_dataloader_idx_in_hooks(tmpdir, multiple_trainloader_mode):
+    """
+    Check the correct dataloader_idx inside hooks
+    """
+
+    class CustomEvalModelTemplate(EvalModelTemplate):
```
Comment (carmocca) [inline tests/trainer/test_dataloaders.py:1170]: Can you inherit from `BoringModel` instead?

**Review by carmocca — COMMENTED** [review]
Great work!

You might also want to add tests to ensure the BC is kept.

File: tests/trainer/test_dataloaders.py (hunk 4 above)
Comment (rohitgr7) [reply tests/trainer/test_dataloaders.py:1170]: Yeah more work 😅

Comment (carmocca) [reply tests/trainer/test_dataloaders.py:1170]: If it's too much, I don't mind leaving it as is. But we want to avoid adding new tests relying on EvalModelTemplate as we want to eventually remove it

Comment (rohitgr7) [reply tests/trainer/test_dataloaders.py]: ok done now :)

File: pytorch_lightning/trainer/training_loop.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py]: done.

**Comment by rohitgr7** [comment]
I have a few more concerns with dataloader_idx w.r.t training. We have on_train_batch_start and some similar training hooks that get dataloader_idx as an argument but for training, I don't see that we need it. Even in the case of multiple train dataloaders it is just 0 since multi-train-dataloader dataflow is different from that of val/test/predict. BTW it defaults to 0 which I think is not correct, should either be removed unless the current behavior is going to be changed in the future or should be marked as None instead of 0.

File: pytorch_lightning/accelerators/accelerator.py [hunk 5]
```diff
@@ -163,12 +166,12 @@ def on_train_start(self) -> None:
...
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/accelerators/accelerator.py:169]: Just for accessibility, although from user's perspective it's BC.

File: pytorch_lightning/accelerators/accelerator.py [hunk 6]
```diff
@@ -453,7 +456,7 @@ def model_sharded_context(self) -> Generator:
             yield
 
     # todo: remove in v1.5
-    def connect_training_type_plugin(self, plugin: TrainingTypePlugin, model: LightningModule) -> None:
+    def connect_training_type_plugin(self, model: LightningModule) -> None:
```
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:459]: you should keep this signature for BC

File: pytorch_lightning/accelerators/accelerator.py [hunk 7]
```diff
@@ -465,10 +468,10 @@ def connect_training_type_plugin(self, plugin: TrainingTypePlugin, model: Lightn
             'Accelerator method `connect_training_type_plugin` was deprecated in v1.3.'
             ' It will be removed in v1.5.'
         )
-        self.setup_training_type_plugin(plugin, model)
+        self.setup_training_type_plugin(model)
 
     # todo: remove in v1.5
-    def connect_precision_plugin(self, plugin: PrecisionPlugin) -> None:
+    def connect_precision_plugin(self) -> None:
```
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:474]: same here

**Review by carmocca — APPROVED** [review]
One BC comment. Other than that LGTM! 💯

File: pytorch_lightning/accelerators/accelerator.py (hunk 6 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/accelerator.py:459]: Isn't this an internal function?

Comment (carmocca) [reply pytorch_lightning/accelerators/accelerator.py:459]: @Borda chose to keep it for BC until 1.5 as indicated by the comment right above.

It should be fine to keep the `plugin` argument for `connect_training_type_plugin` and just not pass it to `setup_training_type_plugin`

There's no need to deprecate the `setup_training_type_plugin` signature change as it is unreleased.

Comment (rohitgr7) [inline pytorch_lightning/accelerators/accelerator.py:459]:
```suggestion
    def connect_training_type_plugin(self, plugin: TrainingTypePlugin, model: LightningModule) -> None:
```

File: pytorch_lightning/accelerators/accelerator.py (hunk 7 above)
Comment (rohitgr7) [inline pytorch_lightning/accelerators/accelerator.py:474]:
```suggestion
    def connect_precision_plugin(self, plugin: PrecisionPlugin) -> None:
```

File: pytorch_lightning/accelerators/accelerator.py (hunk 6 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/accelerator.py]: okay.. Done!

**Comment by rohitgr7** [comment]
Hi everyone, all done from my side here. Please let me know if anything else is required to get this merged. Thanks ✌

**Review by ananthsub — APPROVED** [review]

**Comment by tchaton** [comment]
Hey @rohitgr7,

Mind resolving the conflicts ?

Best,
T.C

File: pytorch_lightning/accelerators/accelerator.py [hunk 8]
```diff
@@ -358,20 +353,23 @@ def setup_optimizers(self, trainer: 'pl.Trainer') -> None:
         self.lr_schedulers = lr_schedulers
         self.optimizer_frequencies = optimizer_frequencies
 
-    def setup_training_type_plugin(self, plugin: TrainingTypePlugin, model: 'pl.LightningModule') -> None:
+    def setup_training_type_plugin(self, model: 'pl.LightningModule') -> None:
         """Attaches the training type plugin to the accelerator."""
-        plugin.setup(model)
+        self.training_type_plugin.setup(model)
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/accelerator.py:358]: do we need to change this in this PR?
same for precision below

File: pytorch_lightning/core/hooks.py [hunk 9]
```diff
@@ -684,22 +688,24 @@ def transfer_batch_to_device(self, batch, device):
             - :meth:`move_data_to_device`
             - :meth:`apply_to_collection`
         """
-        device = device or self.device
```
Comment (awaelchli) [inline pytorch_lightning/core/hooks.py:687]: Why remove this?
This is a useful function to have. Let's keep it optional?

File: tests/deprecated_api/test_remove_1-5.py [hunk 10]
```diff
@@ -123,6 +123,18 @@ def test_v1_5_0_model_checkpoint_period(tmpdir):
         ModelCheckpoint(dirpath=tmpdir, period=1)
 
 
+def test_old_transfer_batch_to_device_hook(tmpdir):
+
+    class OldModel(BoringModel):
+
+        def transfer_batch_to_device(self, batch, device):
+            return super().transfer_batch_to_device(batch, device, None)
```
Comment (awaelchli) [inline tests/deprecated_api/test_remove_1-5.py:131]: but this model only contains one dataloader? Why do we expect a warning?
`validation_step(batch, batch_idx)` doesn't warn about missing dataloader_idx if I have only one dataloader, right?

File: pytorch_lightning/accelerators/accelerator.py (hunk 8 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/accelerator.py:358]: I can't remember why I made this change.. was getting some warning in my new IDE regarding passing the class attributes within class method, so updated this. If not required I will revert it, but still `plugin` argument is not required since it's basically `self.training_type_plugin`.

File: pytorch_lightning/core/hooks.py (hunk 9 above)
Comment (rohitgr7) [reply pytorch_lightning/core/hooks.py:687]: `transfer_batch_to_device` is public api. There the user should get the actual device, otherwise device is always.

File: tests/deprecated_api/test_remove_1-5.py (hunk 10 above)
Comment (rohitgr7) [reply tests/deprecated_api/test_remove_1-5.py:131]: the arguments are consistent here. In case there is only one dataloader then `dataloader_idx is None`. Although there is some inconsistency with this argument in other methods. For eg in `on_train./val/test_batch_start` this is a required argument but in `validation_step` as you said it's not. So I though making it a required argument makes more sense as it makes method inputs same in all the cases so doesn't need to change their API if number of dataloaders are updated.

**Comment by carmocca** [comment]
Needs to remove the following from docs:

**Comment by Borda** [comment]
@rohitgr7 mind check the conflits? :]

**Review by Borda — COMMENTED** [review]
is the `args` >> `kwargs` really needed to be done here as the R title does not mention it, I feel it would be much easier to review split it to two PRs

File: pytorch_lightning/accelerators/accelerator.py [hunk 11]
```diff
@@ -198,71 +201,64 @@ def training_step(
                 - hiddens(:class:`~torch.Tensor`): Passed in if
                   :paramref:`~pytorch_lightning.core.lightning.LightningModule.truncated_bptt_steps` > 0.
         """
-        args[0] = self.to_device(args[0])
+        step_kwargs = self.to_device(step_kwargs)
 
         with self.precision_plugin.train_step_context(), self.training_type_plugin.train_step_context():
-            return self.training_type_plugin.training_step(*args)
+            return self.training_type_plugin.training_step(*step_kwargs.values())
```
Comment (Borda) [inline pytorch_lightning/accelerators/accelerator.py:207]: another n00b question: with dict the order can shuggle also as dist the nb items may yield in order variation...

Comment (Borda) [reply pytorch_lightning/accelerators/accelerator.py:207]: ok, found the `_build_kwargs` but still just thinking if use may accidentally work with this API and mess with passed arguments...

**Review by Borda — APPROVED** [review]
overall the `kwargs` seems much safer than just `args` :]

**Comment by carmocca** [comment]
Unlabeled API/design as this does not modify the API, `dataloader_idx` is there already

**Comment by carmocca** [comment]
cc @awaelchli as accelerator CODEOWNER

File: pytorch_lightning/core/hooks.py [hunk 12]
```diff
@@ -669,6 +669,9 @@ def transfer_batch_to_device(self, batch: Any, device: Optional[torch.device] =
         Note:
             This hook should only transfer the data and not modify it, nor should it move the data to
             any other device than the one passed in as argument (unless you know what you are doing).
+            To check the current state of execution of this hook you can use
+            ``self.trainer.training/testing/validating/predicting`` so that you can
+            add different logic as per your requirement.
```
Comment (awaelchli) [inline pytorch_lightning/core/hooks.py:674]: This sentence could be added to every hook we have in Lightning so I'm wondering how relevant it is here.

File: pytorch_lightning/core/lightning.py [hunk 13]
```diff
@@ -211,9 +214,22 @@ def logger(self):
...
         batch = self.on_before_batch_transfer(batch, dataloader_idx)
-        batch = self.transfer_batch_to_device(batch, device)
+
+        if is_param_in_hook_signature(self.transfer_batch_to_device, 'dataloader_idx'):
+            batch = self.transfer_batch_to_device(batch, device, dataloader_idx)
+        else:
+            warning_cache.warn(
+                "`transfer_batch_to_device` hook signature has changed in v1.4."
+                " `dataloader_idx` parameter has been added to it. Support for"
+                " the old signature will be removed in v1.6", DeprecationWarning
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:229]: will you still show this warning if we are using one dataloader? For one dataloader the dataloader_idx can be optional like it is for our step methods in LightningModule?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/core/lightning.py (hunk 13 above)
Comment (rohitgr7) [reply pytorch_lightning/core/lightning.py:229]: answered here: https://github.com/PyTorchLightning/pytorch-lightning/pull/6241#discussion_r616598429

File: pytorch_lightning/core/hooks.py (hunk 12 above)
Comment (rohitgr7) [reply pytorch_lightning/core/hooks.py:674]: not every. All other hooks are different for train/val/test(for eg on_train_.., on_test...) but this one is a common one for all so there is a possibility that the user needs to have different transformations in train/val/test, that's why added it here for reference.

**Comment by rohitgr7** [comment]
> Unlabeled API/design as this does not modify the API, `dataloader_idx` is there already from #6059

transfer_batch_to_device API is updated.

File: pytorch_lightning/accelerators/accelerator.py (hunk 11 above)
Comment (rohitgr7) [reply pytorch_lightning/accelerators/accelerator.py:207]: sorry I didn't get you. We are using ordered dict so there won't be an issue with order variation.

**Comment by carmocca** [comment]
> transfer_batch_to_device API is updated.

In any case, this was previously approved by William so no need for his explicit approval again.

Feel free to merge

File: pytorch_lightning/core/lightning.py (hunk 13 above)
Comment (kmsquire) [reply pytorch_lightning/core/lightning.py:229]: This suggests that the old signature should still work.  Above, the `dataloader_idx` defaults to `None`, but on the main function in [`core/hooks.py`](https://github.com/PyTorchLightning/pytorch-lightning/pull/6241/files#diff-255f539f021870ad6b64699362bd1b234afe8e606bb4cfd72b54ddba7d0f0326R654), `dataloader_idx` is required (which breaks some of our code that needs to work with earlier versions).
